Add a flag to specify an alternate module root.

creachadair opened the pull request:
Although it is common practice, not all Go repositories have the go.mod file at
the repo root. Add a --moduleRoot flag that allows the caller to specify a path
relative to the repository root, that should be used as the module root for the
purpose of indexing.

creachadair committed:
Add a flag to specify an alternate module root.

Although it is common practice, not all Go repositories have the go.mod file at
the repo root. Add a --moduleRoot flag that allows the caller to specify a path
relative to the repository root, that should be used as the module root for the
purpose of indexing.

creachadair committed:
Bump tool version.

We should do this less manually in the future.

efritz commented on cmd/lsif-go/main.go:
```diff
@@ -96,6 +105,9 @@ func realMain() error {
 
+	// TODO(@creachadair): Packages built with cgo may not include assembly
+	// files.  To index such a package you need to set CGO_ENABLED=0.  Consider
+	// maybe doing this explicitly, always.
```
Oof you're a two spaces guy?

creachadair replied:
I'm not deeply partisan about it. It's just how I learned to type.

efritz replied:
Same here. But I got the double spaces whipped out of me.

creachadair replied:
The idealists who believe computers can automatically get spacing correct without that signal are refreshingly optimistic in their naïveté. Even Dr. Knuth realized it required input from the user, yet there are some communities you cannot blast off their high horses. (I've removed the spaces, since I really don't care that much especially in plain text comments—but I reserve the right to continue laughing and shaking my head at abject prescriptivists)

efritz commented on internal/gomod/module.go:
```diff
@@ -60,7 +60,7 @@ func ListModules(projectRoot string) (string, map[string]string, error) {
 	}
 
-	out, err := run(projectRoot, "go", "list", "-m", "all")
+	out, err := run(projectRoot, "go", "list", "-mod=readonly", "-m", "all")
```
Why this addition? Will this require that mod already be fetched?

creachadair replied:
Without this, a package with vendored dependencies can't be listed. It should be harmless for any package without vendored dependencies. It does require that the dependencies be already loaded, but if that isn't the case then the indexer mutates the module cache which seems wrong.

We could add a `go mod download` step first, though.

efritz replied:
To the tool or to the CI step?

creachadair replied:
Either or both.

creachadair replied:
For further context: I think it was a mistake for us to write the indexer in such a way that it will touch the `go` tool for anything other than reading metadata. Indexing should either be done _during_ a build, so that the dependencies for each package being indexed are already populated, or _after_ a build so that the dependencies of the whole module are populated.

But that ship may have already sailed, in which case doing it here may be the least-pessimal workaround.

creachadair replied:
I added a call to `go mod download`. That preserves the existing behaviour for modules without vendored dependencies.

efritz commented on cmd/lsif-go/main.go:
```diff
@@ -47,6 +50,7 @@ func realMain() error {
 	app.Flag("out", "The output file.").Short('o').Default("dump.lsif").StringVar(&outFile)
 	app.Flag("moduleVersion", "Specifies the version of the module defined by this project.").PlaceHolder("version").StringVar(&moduleVersion)
 	app.Flag("repositoryRoot", "Specifies the path of the current repository (inferred automatically via git).").PlaceHolder("root").StringVar(&repositoryRoot)
+	app.Flag("moduleRoot", "Specifies the module root directory relative to the repository").Default(".").StringVar(&moduleRoot)
```
What is the difference between doing this and running the indexer in the correct root instead?

creachadair replied:
It doesn't work: If you run it from a subdirectory of the repository, the tool uses `git` to find the real root and looks for the module there. Similarly if you set `--repositoryRoot`, as it uses that only as the starting point. 

This whole interface is kind of bad, but I didn't want to break existing use cases. Properly we shouldn't be using `git` to derive the module path, but I wanted to avoid having to update all our existing config files.

efritz replied:
Would now be the item to update the interface?

creachadair committed:
Clarify the TODO comment.

efritz commented on cmd/lsif-go/main.go:
```diff
@@ -96,6 +105,9 @@ func realMain() error {
 		Args:    os.Args[1:],
 	}
 
+	// TODO(@creachadair): With cgo enabled, the indexer cannot handle packages
```
😛

creachadair committed:
Remove a space.

creachadair committed:
Remove another space.

creachadair committed:
Ensure modules are downloaded before listing.

To handle vendored dependencies, we use go list -mod=readonly.
That, however, requires we explicitly fetch any dependencies before listing to
preserve the existing behaviour.

Ideally we we would not modify the environment, but this will handle the vendor
case without breaking existing use.

efritz commented:
I'm assuming that this works fine on the repos we're currently indexing in CI.

creachadair commented:
> I'm assuming that this works fine on the repos we're currently indexing in CI.

I tested it against a few locally and it works fine, but I haven't actually tried it in GitHub Actions yet. I think I may need to update the actions repo to pick up the new version. When I do, I'll check it against some live ones and roll it back if I missed something.
